### PR TITLE
Integrates jest-puppeteer for UI testing

### LINKED ISSUE #503: Adding Test suite
Integrating the jest-puppeteer for testing and writing tests in a similar way to that implemented in Image-Sequencer. Jest-puppeteer facilitates Keyboard Typing and Mouse simulation so would be best suitable here as well.  
Adding test For Bold Formatting 
Adding test For Italic Formatting

Jest-puppeteer was previously used in the Image-Sequencer project and they work flawlessly. So we can implement it here as well

### PR DESCRIPTION
**Pull request by shreyaa-sharmaa**
Fixes #503 
Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt jasmine`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled
* [x] PR body includes `fixes #0000`-style reference to original issue #
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

We have a loose schedule of reviewing and pulling in changes every Tuesday and Friday, and publishing changes on Fridays. 

Thanks!

### CODE REVIEW

**Comment by shreyaa-sharmaa** [comment]
It doesn't work without adding args: ['--no-sandbox'] in the config file. (Found this as a solution for the errors) . The only relevant piece of info I've found so far is 
```When running headless Chrome in a container without a defined user, the chromeOptions environment property needs a --no-sandbox arg (in addition to the other headless args), or Chrome won't be able to startup.```
Are there any negative impacts for using this? Is it just a workaround or a required necessity? Is there a better approach for this?

**Comment by shreyaa-sharmaa** [comment]
The jest test can be run using `npm run test-ui`. If you face any error while running these tests:
1. Make sure you're not running them as a root user. If running them as a root user include  `args: ['--no-sandbox']` in `jest-puppeteer.config.js` (not recommended).
2. Make sure you have all the required dependencies installed.  
For Debian systems 
```
sudo apt-get install gconf-service libasound2 libatk1.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libexpat1 libfontconfig1 libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libnspr4 libpango-1.0-0 libpangocairo-1.0-0 libstdc++6 libx11-6 libx11-xcb1 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxext6 libxfixes3 libxi6 libxrandr2 libxrender1 libxss1 libxtst6 ca-certificates fonts-liberation libappindicator1 libnss3 lsb-release xdg-utils wget
```
3. Upgrade your kernel.

**Comment by jywarren** [comment]
This is looking awesome. Can you include your docs (as written above) in the README as a part of this PR? 

This is really cool! Just noting we can potentially merge this PR without making the UI tests required, in order to allow people to try it out and to refine it a bit. Then we can open a follow-up issue to "turn it on" for Travis and make them required. But, we don't have to do it this way, it's just an option to consider!

Thanks @shreyaa-sharmaa this is going to be EPICALLY useful for any new features we add, as every PR can begin to have thorough UI tests. Please consider including it in either the next release in #532 or in an upcoming release! 🎉 🙌

**Comment by jywarren** [comment]
@Shulammite-Aso @keshav234156 @NitinBhasneria this is really cool! Give it a try?

And one more thought - if you need special configurations to run this, perhaps adding lines to our [Travis configuration](https://github.com/publiclab/PublicLab.Editor/blob/main/.travis.yml) to ensure it will run in /that/ environment, will help to demonstrate the prerequisites needed to run it, all in the same PR?

Here's an example complex travis.yml with prerequisites: https://github.com/publiclab/image-sequencer/blob/main/.travis.yml

And here are some docs! https://docs.travis-ci.com/user/installing-dependencies/

Great work! 💯

**Comment by shreyaa-sharmaa** [comment]
Thanks for the input @jywarren . I completely agree with the idea of not making the UI tests required as of right now. The only reason I opened it as a draft PR was because I was writing a few tests and wanted to make sure that the config file meets all the requirements before I open the PR. Since we faced a few bugs with the formatting options like bolding the text, and have been working on implementing the new features, I too believe that this would be a great asset for thorough UI testing and would love to see it in the next release.  I'll go through the docs that you shared and get back here as soon as possible. Also I'll try to add in a few tests asap. I'm working on them right now.

**Comment by jywarren** [comment]
Hi @shreyaa-sharmaa can you link us to where the tests are being run in Travis, if they are yet? Or if not, can you paste in here (or quote from gist.github.com) a sample test log? Thanks!!

**Comment by keshav234156** [comment]
Please add the test command in travis.yml,so that it  an run on Travis

**Comment by shreyaa-sharmaa** [comment]
@keshav234156 do I need to make any changes other than that in travis.yml? Simply adding the `npm run test-ui` didin't work. Neither did adding the required dependencies.

**Comment by shreyaa-sharmaa** [comment]
It's done!
Screenshot from 2020-06-28 15-07-04

Thanks everyone! Please review and let me know if any changes are required.

**Comment by jywarren** [comment]
Wow, this is fantastic! Have you tested introducing a test error to be sure Travis fails too? Just this last test -- we should see a little red X appear -- and then let's fix it again and merge this!

**Comment by jywarren** [comment]
Actually, because I know other PRs are waiting on this, would you mind @shreyaa-sharmaa creating a PR that breaks tests, just to confirm, and linking here? Then I can merge this to unblock others. FANTASTIC WORK! This is going to be huge for complex features we need to protect, and for debugging delicate things like https://github.com/jywarren/woofmark/pull/4#issuecomment-648808935

THANKS!! 🎉 🎉

File: README.md [hunk 1]
```diff
@@ -320,14 +320,26 @@ For additional support, join the Public Lab website and mailing list at http://p
 
 ## Testing
 
-Automated tests are an essential way to ensure that new changes don't break existing functionality, and can help you be confident that your code is ready to be merged in. We use Jasmine for testing: https://jasmine.github.io/2.4/introduction.html
+Automated tests are an essential way to ensure that new changes don't break existing functionality, and can help you be confident that your code is ready to be merged in. We use Jasmine for testing: https://jasmine.github.io/2.4/introduction.html. The UI tests are written using jest and puppeteer.
```
Comment (jywarren) [inline README.md:323]: THANK YOU for this excellent documentation!

**Comment by jywarren** [comment]
Ah, actually look - it failed here after updating!! https://github.com/publiclab/PublicLab.Editor/pull/543

so we should be all good! ✅ 🎉 🙌

**Pull request merged**